### ISSUE TITLE
64 bit shift is broken in nightly build for MIPS

### ISSUE THREAD

**Issue by mcroomp** [rust-lang/rust#116177]
I tried this code:

```rust
const SHIFTVALUE : i64 = 10000000000i64;

fn main() {

    fn shiftleft(a : i64, b : i64) -> i64 {
        a << b
    }

    assert_eq!(shiftleft(black_box(SHIFTVALUE), black_box(2)), SHIFTVALUE << 2);
}

```
Nightly fails running on MIPS emulator 

assertion `left == right` failed
  left: 35705032704
 right: 40000000000

Stable succeeds

Looks like the top and bottom dwords are mixed up

target is mips-unknown-linux-musl
emulator is qemu-mips-static

**Comment by nikic** [comment]
This is probably related to https://github.com/llvm/llvm-project/issues/64794.

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23116177.2064.20bit.20shift.20is.20broken.20in.20nightly.20build.20for.20MIPS)).

@rustbot label -I-prioritize +P-medium

**Comment by saethlin** [comment]
This can be provoked on misp64 as well:
```rust
use std::hint::black_box;

const SHIFTVALUE: i128 = u64::MAX as i128;

fn main() {
    fn shiftleft(a: i128, b: i128) -> i128 {
        a << b
    }

    assert_eq!(
        shiftleft(black_box(SHIFTVALUE), black_box(2)),
        SHIFTVALUE << 2
    );
}
```

**Comment by matoro** [comment]
Is this what's causing these build failures when upgrading from 1.72.0 to 1.73.0?  Just checking before I open a new bug.

```
thread 'rustc' panicked at compiler/rustc_mir_build/src/build/mod.rs:1004:13:
apfloat::ieee::Double gave different result for `3.321928094887362`: 3.3219319068173765(0x400a9351091a8e71) vs Rust's 3.3219280948873622(0x400a934f0979a371)
```

I get a bunch of these on slightly different floating-point values on mips64.

**Comment by saethlin** [comment]
Floats are hell, and also this landed with 1.73: https://github.com/rust-lang/rust/pull/113843

A new issue is a good idea.

**Comment by CryZe** [comment]
Yeah I've also been getting lots of different floating point related issues on 1.73 on MIPS. I too would've thought they are related to the shifts, but the new apfloat code could also be the reason.

**Comment by matoro** [comment]
I applied the patch from https://github.com/llvm/llvm-project/issues/64794#issuecomment-1743354405 and confirmed that it fixes the 1.73.0 bootstrap problem for me.  I also tried the snippet from @saethlin 's comment above and it did not assert after applying that patch.

**Comment by Cyanoxygen** [comment]
It happened with `{u,i}128`, too. We stripped down the PoC to a few lines of LLVM IR code, and we reported to the [upstream](https://github.com/llvm/llvm-project/issues/71142).  
Turned out that it is exactly the same problem here.
